compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -111,7 +111,18 @@ impl FromInternal<(TokenStream, &mut Rustc<'_, '_>)> for Vec<TokenTree<TokenStre
                     }));
                     continue;
                 }
-                tokenstream::TokenTree::Token(token, spacing) => (token, spacing == Joint),
+                tokenstream::TokenTree::Token(token, fby) => {
+                    // Do not be tempted to check here that the `fby` values
+                    // are correct w.r.t. the token stream (e.g. that
+                    // `FollowedBy::Punct` is actually followed by a `Punct`
```

Comment:
This checking is exactly what we need to do here to do the censoring without threading `FollowedBy::{Punct,Other}` through the rest of the compiler.

The fact that it causes issues means that there's an issue elsewhere.